server/handler.go

```diff
@@ -95,23 +95,28 @@ func (h *Handler) ComInitDB(c *mysql.Conn, schemaName string) error {
 	return h.sm.SetDB(c, schemaName)
 }
 
+// ComPrepare parses, partially analyzes, and caches a prepared statement's plan
+// with the given [c.ConnectionID].
 func (h *Handler) ComPrepare(c *mysql.Conn, query string) ([]*query.Field, error) {
 	ctx, err := h.sm.NewContextWithQuery(c, query)
 	if err != nil {
 		return nil, err
 	}
-	schema, err := h.e.AnalyzeQuery(ctx, query)
+
+	analyzed, err := h.e.PrepareQuery(ctx, query)
 	if err != nil {
 		return nil, err
 	}
-	if sql.IsOkResultSchema(schema) {
+
+	h.e.CachePreparedStmt(c.ConnectionID, analyzed, query)
```

Comment:
e.PrepareQuery should handle this caching itself, no reason for this to be separate.